api/v1alpha1/addon_types.go

```diff
@@ -94,69 +97,79 @@ type NFD struct {
 type ClusterAutoscaler struct {
 	// Addon strategy used to deploy cluster-autoscaler to the management cluster
 	// targeting the workload cluster.
+	// +kubebuilder:validation:Required
 	// +kubebuilder:validation:Enum=ClusterResourceSet;HelmAddon
 	Strategy AddonStrategy `json:"strategy"`
 }
 
 type DefaultStorage struct {
 	// Name of the CSI Provider for the default storage class.
+	// +kubebuilder:validation:Required
 	// +kubebuilder:validation:Enum=aws-ebs;nutanix
 	ProviderName string `json:"providerName"`
+
 	// Name of storage class config in any of the provider objects.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
 	StorageClassConfigName string `json:"storageClassConfigName"`
 }
 
 type CSI struct {
-	// +optional
+	// +kubebuilder:validation:Optional
 	Providers []CSIProvider `json:"providers,omitempty"`
-	// +optional
+
+	// +kubebuilder:validation:Optional
 	DefaultStorage *DefaultStorage `json:"defaultStorage,omitempty"`
 }
 
 type CSIProvider struct {
 	// Name of the CSI Provider.
+	// +kubebuilder:validation:Required
 	// +kubebuilder:validation:Enum=aws-ebs;nutanix
 	Name string `json:"name"`
 
-	// +optional
+	// +kubebuilder:validation:Optional
 	StorageClassConfig []StorageClassConfig `json:"storageClassConfig,omitempty"`
 
 	// Addon strategy used to deploy the CSI provider to the workload cluster.
+	// +kubebuilder:validation:Required
 	// +kubebuilder:validation:Enum=ClusterResourceSet;HelmAddon
 	Strategy AddonStrategy `json:"strategy"`
 
 	// The reference to any secret used by the CSI Provider.
-	// +optional
-	Credentials *corev1.LocalObjectReference `json:"credentials,omitempty"`
+	// +kubebuilder:validation:Optional
+	Credentials *LocalObjectReference `json:"credentials,omitempty"`
 }
 
 type StorageClassConfig struct {
 	// Name of storage class config.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
 	Name string `json:"name"`
 
 	// Parameters passed into the storage class object.
-	// +optional
+	// +kubebuilder:validation:Optional
```

Comment:
Per https://book.kubebuilder.io/reference/markers.html
<pre>
If you’re using controller-gen only then they’re redundant, but if you’re using other generators or you want developers that need to build their own clients for your API, you’ll want to also include +optional.
</pre>
do we expect anyone building their own clients? ofcourse following also says
<pre>
The most reliable way in 1.x to get +optional is omitempty.
</pre>
which we do.

Reply:
I don't expect any clients to be built from this, especially as these are not CRDs to have a client built for.

> The most reliable way in 1.x to get +optional is omitempty.

We're on 3.x though 😉

Reply:
I think it's ok to leave with `+kubebuilder:validation:Optional` as a companion of `+kubebuilder:validation:Required` here. Are you OK with that @deepakm-ntnx ? We can add `+optional` back later if we find we need to, but I don't expect we will.

Reply:
sure makes sense. thanks